tstest/clock.go

```diff
@@ -4,57 +4,495 @@
 package tstest
 
 import (
+	"container/heap"
 	"sync"
 	"time"
+
+	"tailscale.com/tstime"
+	"tailscale.com/util/mak"
 )
 
+// ClockBuilder is used to configure the initial settings for a Clock. Once
+// the settings are configured as desired, call ClockBuilder.Finish to get the
+// resulting Clock. ClockBuilder can only be used to create a single Clock.
+type ClockBuilder struct {
+	c *Clock
+}
+
+func NewClockBuilder() *ClockBuilder {
+	cb := &ClockBuilder{
+		c: new(Clock),
+	}
+	return cb
+}
+
+func (cb *ClockBuilder) Start(t time.Time) *ClockBuilder {
+	cb.c.start = t
+	return cb
+}
+
+func (cb *ClockBuilder) Step(d time.Duration) *ClockBuilder {
+	cb.c.step = d
+	return cb
+}
+
+func (cb *ClockBuilder) TimerChannelSize(n int) *ClockBuilder {
+	cb.c.timerChannelSize = n
+	return cb
+}
+
+func (cb *ClockBuilder) Finish() *Clock {
+	c := cb.c
+	cb.c = nil
+	c.init()
+	return c
+}
+
+// NewClock returns a Clock initialized with default settings (equivalent to
+// NewClockBuilder().Finish()).
+func NewClock() *Clock {
+	c := new(Clock)
+	c.init()
+	return c
+}
+
 // Clock is a testing clock that advances every time its Now method is
-// called, beginning at Start.
-//
-// The zero value starts virtual time at an arbitrary value recorded
-// in Start on the first call to Now, and time never advances.
+// called, beginning at its start time. If no start time is specified using
+// ClockBuilder, an arbitrary start time will be selected when the Clock is
+// created and can be retrieved by calling Clock.Start().
 type Clock struct {
-	// Start is the first value returned by Now.
-	Start time.Time
-	// Step is how much to advance with each Now call.
-	Step time.Duration
-	// Present is the time that the next Now call will receive.
-	Present time.Time
+	// start is the first value returned by Now. It must not be modified after
+	// init() is called.
+	start time.Time
+
+	mu sync.Mutex
 
-	sync.Mutex
+	// step is how much to advance with each Now call.
+	step time.Duration
+	// present is the last value returned by Now (and will be returned again by
+	// ConstantNow).
+	present time.Time
+	// skipStep indicates that the next call to Now should not add step to
+	// present. This occurs after initialization and after Advance.
+	skipStep bool
+	// timerChannelSize is the buffer size to use for channels created by
+	// NewTimer and NewTicker.
+	timerChannelSize int
```

Comment:
Is it important for the caller to care about this? Since this is for tests only, any downside to making the channel absurdly large (e.g. 4k slots) ? Or is it important that callers can set up loss of notifications?

Reply:
I don't know, I feel like loss by default seems good but sometimes loss will be bad. I think we kind of need to see what the tests end up looking like that need protection from loss in order to figure it out.